pvfs-1.5.0/docs/pvfs_lseek.3
.\" Hey Emacs! This file is -*- nroff -*- source.
.\"
.\" This manpage is copyright (c) 1997 Clemson University.
.\"
.\" Written by Rob Ross and Matt Cettei.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\"
.\" The author(s) assume no responsibility for errors or omissions, or
.\" for damages resulting from the use of the information contained herein.
.\"
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" Contact:  Rob Ross    rbross@parl.eng.clemson.edu
.\"           Matt Cettei mcettei@parl.eng.clemson.edu
.\" 
.TH PVFS_LSEEK 3 "2 December 1997" "PVFS calls"
.SH NAME
pvfs_lseek \- reposition read/write file offset
.SH SYNOPSIS
.B #include <unistd.h>
.sp
.BI "off_t pvfs_lseek(int " fildes ", off_t " offset ", int " where );
.SH DESCRIPTION
The
.B pvfs_lseek
function repositions the offset of the file descriptor
.I fildes
to the argument
.I offset
according to the directive
.IR  where .
The argument
.I fildes
must be an open file descriptor.
.B Lseek
repositions the file pointer
.I fildes
as follows:
.RS
.sp
If
.I where
is
.BR SEEK_SET ,
the offset is set to
.I offset
bytes.

If
.I where
is
.BR SEEK_CUR ,
the offset is set to its current location plus
.I offset
bytes.

If
.I where
is
.BR SEEK_END ,
the offset is set to the size of the file plus
.I offset
bytes.
.RE

The
.B pvfs_lseek
function allows the file offset to be set beyond the end of the existing
end-of-file of the file. If data is later written at this point, subsequent
reads of the data in the gap return bytes of zeros (until data is actually
written into the gap).

Note that there is some added complexity in PVFS with regards to seeking.
Specifically, the locations specified via
.B pvfs_lseek
refer to a location in the current
.I partition
for the file descriptor.  This means that some extra thought should be
taken when determining seek offsets.

.SH "RETURN VALUES"
Upon successful completion,
.B pvfs_lseek
returns the resulting offset location in the partition as measured in
bytes from the beginning of the partition.  Otherwise, a value of \-1 is
returned and
.I errno
is set to indicate the error.  See the 
.BR lseek "(2)
manpage for 
.I errno
values in the event of an error.
.SH "CONFORMING TO"
Nothing
.SH "SEE ALSO"
.BR dup "(2), " open "(2), " fseek (3)
.SH AUTHOR
Rob Ross, Clemson University

Based on
.B ioctl
manpage from BSD manpages.
.SH BUGS
I dumped the use of whence; it's silly.  Yes, there might be bugs.


pvfs-1.5.0/lib/pvfs_lseek.c
/* pvfs_lseek.c copyright (c) 1995 Clemson University, all rights reserved.
	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License as published
	by the Free Software Foundation; either version 2 of the License, or
	any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of 
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
	General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
	Contact:  Rob Ross  	rbross@parl.eng.clemson.edu
				 Matt Cettei mcettei@parl.eng.clemson.edu

 */

/* PVFS_LSEEK.C - parallel seek call
 *
 * pvfs_lseek() moves the read-write file offset on a parallel file in a 
 * manner similar to the unix lseek() call.  It accepts 3 parameters: a
 * parallel file descriptor, an offset value, and a "whence" parameter 
 * used to specify how to interpret the offset.
 *
 * whence may have 1 of two values, SEEK_SET or SEEK_CUR.  SEEK_SET
 * specifies to set the offset to the value given.  SEEK_CUR specifies that
 * the offset value should be added to the current offset.  Note that the
 * offset value may be negative.
 *
 * update: whence may now have the value of SEEK_END.  I finally got around
 * to fixing it.
 *
 * Upon successful completion, pvfs_lseek() returns the current offset, 
 * measured in bytes from the beginning of the file.  If it fails, 
 * -1 is returned and the offset is not changed.
 */

static char pvfs_lseek_c[] = "$Header: /projects/cvsroot/pvfs/lib/pvfs_lseek.c,v 1.4 2000/10/30 14:44:49 rbross Exp $";

/* UNIX INCLUDE FILES */
#include <errno.h>
#include <unistd.h>

/* PVFS INCLUDE FILES */
#include <lib.h>

/* GLOBALS */
extern int errno;
extern fdesc_p pfds[];

/* FUNCTIONS */
int pvfs_lseek(int fd, int off, int whence)
{
	int i, val;
	struct stat filestat;

	if (fd < 0 || fd >= NR_OPEN || (pfds[fd] && pfds[fd]->fs == FS_RESV)) {
		errno = EBADF;
		return(-1);
	}  

	/* we're handling directories within PVFS now, so don't call lseek() */
	/* for them. */
	if (!pfds[fd] || pfds[fd]->fs == FS_UNIX) {
		if (pfds[fd]) /* gotta keep our info up to date */ {
			return(pfds[fd]->fd.off = lseek(fd, off, whence));
		}
		else {
			return lseek(fd, off, whence);
		}
	}

	switch(whence) {
		case SEEK_SET:
			if (off >= 0) return(pfds[fd]->fd.off = off);
			errno = EINVAL;
			return(-1);
		case SEEK_CUR:
			if (pfds[fd]->fd.off + off >= 0) return(pfds[fd]->fd.off += off);
			errno = EINVAL;
			return(-1);
		case SEEK_END:
			/* for directories, this doesn't work yet */
			if (pfds[fd]->fs == FS_PDIR) {
				errno = EINVAL;
				return(-1);
			}
			/* find the actual end of the file */
			if ((i = pvfs_fstat(fd, &filestat)) < 0) {
				PERROR("Getting file size");
				return(-1);
			}
			if (off + filestat.st_size >= 0)
				return(pfds[fd]->fd.off = off + filestat.st_size);
		/* let error here and default drop through...behavior is the same */
	}
	errno = EINVAL;
	return(-1);
} /* end of PVFS_LSEEK() */
